imagemanager/imagemanager.go

```diff
@@ -0,0 +1,1076 @@
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"time"
+
+	"github.com/aoscloud/aos_common/aoserrors"
+	"github.com/aoscloud/aos_common/aostypes"
+	"github.com/aoscloud/aos_common/api/cloudprotocol"
+	"github.com/aoscloud/aos_common/image"
+	"github.com/aoscloud/aos_common/spaceallocator"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/aoscloud/aos_communicationmanager/config"
+	"github.com/aoscloud/aos_communicationmanager/fcrypt"
+	"github.com/aoscloud/aos_communicationmanager/unitstatushandler"
+)
+
+/***********************************************************************************************************************
+ * Consts
+ **********************************************************************************************************************/
+
+const (
+	decryptedFileExt = ".dec"
+
+	fileScheme = "file"
+	httpScheme = "http"
+
+	blobsFolder = "blobs"
+)
+
+/***********************************************************************************************************************
+ * Types
+ **********************************************************************************************************************/
+
+// Storage provides API to create, remove or access information from DB.
+type Storage interface {
+	GetServiceVersions(serviceID string) (services []ServiceInfo, err error)
+	GetServicesInfo() ([]ServiceInfo, error)
+	GetLayersInfo() ([]LayerInfo, error)
+	GetServiceInfo(serviceID string) (ServiceInfo, error)
+	GetLayerInfo(digest string) (LayerInfo, error)
+	AddLayer(layer LayerInfo) error
+	AddService(service ServiceInfo) error
+	SetLayerCached(digest string, cached bool) error
+	SetServiceCached(serviceID string, cached bool) error
+	RemoveService(serviceID string, aosVersion uint64) error
+	RemoveLayer(digest string) error
+}
+
+// CryptoContext interface to access crypto functions.
+type CryptoContext interface {
+	ImportSessionKey(keyInfo fcrypt.CryptoSessionKeyInfo) (symmetricContext fcrypt.SymmetricContextInterface, err error)
+	CreateSignContext() (signContext fcrypt.SignContextInterface, err error)
+}
+
+// Imagemanager image manager instance.
+type Imagemanager struct {
+	server                 *http.Server
+	layersDir              string
+	servicesDir            string
+	tmpDir                 string
+	storage                Storage
+	cryptoContext          CryptoContext
+	serviceAllocator       spaceallocator.Allocator
+	layerAllocator         spaceallocator.Allocator
+	tmpAllocator           spaceallocator.Allocator
+	serviceTTLDays         uint64
+	layerTTLDays           uint64
+	validateTTLStopChannel chan struct{}
+}
+
+// ServiceInfo service information.
+type ServiceInfo struct {
+	ID         string
+	AosVersion uint64
+	LocalURL   string
+	RemoteURL  string
+	Path       string
+	Size       uint64
+	Timestamp  time.Time
+	Cached     bool
+	Config     aostypes.ServiceConfig
+	Layers     []string
+}
+
+// LayerInfo service information.
+type LayerInfo struct {
```

Comment:
I have added similar structs in aostypes. please check MP 60

Reply:
discussed not the same